packages/gatsby-plugin-sharp/src/scheduler.js

```diff
@@ -1,88 +1,140 @@
 
-const setJobToProcess = (toProcess, job, deferred) => {
-  const inputFileKey = getFileKey(job.inputPath)
-  const outputFileKey = getFileKey(job.outputPath)
-  const jobPath = `["${inputFileKey}"].["${outputFileKey}"]`
-
-  // Check if the job has already been queued. If it has, there's nothing
-  // to do, return.
-  if (_.has(toProcess, jobPath)) {
-    return { existingPromise: _.get(toProcess, `${jobPath}.deferred.promise`) }
+// node 8 doesn't support promise.finally, we extract this function to re-use it inside then & catch
+const cleanupJob = (job, boundActionCreators) => {
+  if (bar) {
+    bar.tick(job.task.args.operations.length)
   }
 
-  // Check if the output file already exists so we don't redo work.
-  if (existsSync(job.outputPath)) {
-    return { existingPromise: Promise.resolve(job) }
-  }
+  imagesFinished += job.task.args.operations.length
 
-  let isQueued = false
-  if (toProcess[inputFileKey]) {
-    isQueued = true
+  if (imagesFinished === imagesToProcess) {
+    if (bar) {
+      bar.done()
+      bar = null
+    }
+    imagesToProcess = 0
+    imagesFinished = 0
   }
 
-  _.set(toProcess, jobPath, {
-    job: job,
-    deferred,
-  })
-
-  return { isQueued }
+  boundActionCreators.endJob({ id: job.id }, { name: `gatsby-plugin-sharp` })
 }
 
+const executeJobs = _.throttle(
+  (pluginOptions, boundActionCreators) => {
+    toProcess.forEach(job => {
+      const { task } = job
+      toProcess.delete(task.inputPaths[0])
+
+      try {
+        worker
+          .IMAGE_PROCESSING(task.inputPaths, task.outputDir, task.args)
+          .then(() => {
+            job.deferred.resolve()
+            cleanupJob(job, boundActionCreators)
+          })
+          .catch(err => {
+            job.deferred.reject(err)
+            cleanupJob(job, boundActionCreators)
+          })
+      } catch (err) {
+        job.deferred.reject(err)
+        cleanupJob(job, boundActionCreators)
+      }
+    })
+  },
+  1000,
+  { leading: false }
+)
+
 const scheduleJob = async (
   job,
   boundActionCreators,
-  pluginOptions,
   reporter,
-  reportStatus = true
+  reportStatus = true,
+  pluginOptions
 ) => {
-  // deferred naming comes from https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Deferred
-  let deferred = {}
-  deferred.promise = new Promise((resolve, reject) => {
-    deferred.resolve = resolve
-    deferred.reject = reject
-  })
-
-  const { existingPromise, isQueued } = setJobToProcess(
-    toProcess,
-    job,
-    deferred
-  )
-  if (existingPromise) {
-    return existingPromise
-  }
+  const isQueued = toProcess.has(job.inputPath)
+  let scheduledPromise
 
-  if (totalJobs === 0) {
+  if (reportStatus && !bar) {
     bar = createProgress(`Generating image thumbnails`, reporter)
     bar.start()
   }
-  totalJobs += 1
 
-  if (!isQueued) {
-    // Create image job
+  // Check if the output file already exists so we don't redo work.
+  if (existsSync(job.outputPath)) {
```

Comment:
Add a TODO to remove once we hoist the check for if a job needs to be run or not to the job manager

```suggestion
  // TODO: Remove this once we hoist the check for if a job needs to be run or not to the job manager
  if (existsSync(job.outputPath)) {
```